### PR TITLE
[CARBONDATA-3648] Support Alter Table Compaction Level Threshold

### PR DESCRIPTION
**Pull request by marchpure**
### Why is this PR needed?
(1) The Alter Table sould support Compaction Level Threshold.
(2) The upper limit of Compaction Level Threshold is 100, which is too small to meet the scenario with massive small files
(3) There is a bug to limit alter table in windows env.
 
 ### What changes were proposed in this PR?
(1) AlterTableUtil support alter Compaction_Level_Threshold
(2) CarbonProperties increases the upper limit of Compaction_Level_Threshold to 10000 from 100.
(3) Fix the bug which limit alter table in windows env
    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1380/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1390/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1402/

File: core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java [hunk 1]
```diff
@@ -81,7 +81,7 @@ public static String getFolderContainingFile(String carbonFilePath) {
     int lastIndex = carbonFilePath.lastIndexOf('/');
     // below code for handling windows environment
     if (-1 == lastIndex) {
-      lastIndex = carbonFilePath.lastIndexOf(CarbonCommonConstants.FILE_SEPARATOR);
+      lastIndex = carbonFilePath.lastIndexOf(CarbonCommonConstants.WINDOWS_FILE_SEPARATOR);
```
Comment (zzcclp) [inline core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java]: why it need to use windows file separator?

Comment (marchpure) [reply core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java]: carbonFilePath is in type of '/root/schema', while it is in type of 'D:\schema'. The separator is different, dividing into FILE_SEPARATOR ('/') and WINDOWS_FILE_SEPARATOR ('\'). 

In the Windows Env.  The variant 'lastIndex' is always -1, carbonFilePath.substring(0, lastIndex) will be throw SubStringIndexOutOfBound Exception.

Comment (zzcclp) [reply core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java]: in the linux env, is lastIndex right if use 'WINDOWS_FILE_SEPARATOR'?
it better set the right separator according to system env, for example: use java.io.File.separator

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1385/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1407/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1395/

**Comment by marchpure** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1386/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1396/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1408/

File: core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java (hunk 1 above)
Comment (marchpure) [reply core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java]: Yes, java.io.File.separator is more elegant. I have modified code to use java.io.File.separator as you suggested.

**Comment by ydvpankaj97** [comment]
retest this please

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 2]
```diff
@@ -802,6 +802,17 @@ private CarbonCommonConstants() {
    */
   public static final String DEFAULT_SEGMENT_LEVEL_THRESHOLD = "4,3";
 
+
+  /**
+   * Min value of Number of segments to be merged in levels
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: What does `in levels` mean? Can you rephrase it?

File: integration/spark2/src/main/scala/org/apache/spark/util/AlterTableUtil.scala [hunk 3]
```diff
@@ -533,6 +533,7 @@ object AlterTableUtil {
       "COMMENT",
       "COLUMN_META_CACHE",
       "CACHE_LEVEL",
+      "COMPACTION_LEVEL_THRESHOLD",
```
Comment (ajantha-bhat) [inline integration/spark2/src/main/scala/org/apache/spark/util/AlterTableUtil.scala]: Here it is added as a table properties right ? Need to add validation also ?
I guess now it will accept any value.

File: core/src/main/java/org/apache/carbondata/core/util/CarbonProperties.java [hunk 4]
```diff
@@ -983,8 +983,11 @@ public void print() {
     for (String levelSize : levels) {
       try {
         int size = Integer.parseInt(levelSize.trim());
-        if (validate(size, 100, 0, -1) < 0) {
+        if (validate(size, CarbonCommonConstants.COMPACTION_SEGMENT_LEVEL_MAX,
+            CarbonCommonConstants.COMPACTION_SEGMENT_LEVEL_MIN, -1) < 0) {
           // if given size is out of boundary then take default value for all levels.
+          LOGGER.warn("The COMPACTION_LEVEL_THRESHOLD is out of boundary, "
```
Comment (ajantha-bhat) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonProperties.java]: print the default value also, similar log like line : 995

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/alterTable/TestAlterTableCompactionLevel.scala [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  }
+
+  override def beforeAll {
+    // drop table
+    dropTable
+    // create table
+    sql("create table alter_compaction_level(c1 String) stored by 'carbondata'")
+  }
+
+  test("validate alter compaction level") {
```
Comment (ajantha-bhat) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/alterTable/TestAlterTableCompactionLevel.scala]: can you add a test case for more than 10000 threshold?

As now there is no validation is present, it will pass !!

**Comment by marchpure** [comment]
retest this please

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java (hunk 2 above)
Comment (marchpure) [reply core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: solved

File: integration/spark2/src/main/scala/org/apache/spark/util/AlterTableUtil.scala (hunk 3 above)
Comment (marchpure) [reply integration/spark2/src/main/scala/org/apache/spark/util/AlterTableUtil.scala]: solved

File: core/src/main/java/org/apache/carbondata/core/util/CarbonProperties.java (hunk 4 above)
Comment (marchpure) [reply core/src/main/java/org/apache/carbondata/core/util/CarbonProperties.java]: solved

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/alterTable/TestAlterTableCompactionLevel.scala (hunk 5 above)
Comment (marchpure) [reply integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/alterTable/TestAlterTableCompactionLevel.scala]: solved

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1743/

**Comment by marchpure** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1745/

**Comment by jackylk** [comment]
LGTM
